flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/SourceWatermarkITCase.scala

```diff
@@ -305,4 +306,56 @@ class SourceWatermarkITCase extends StreamingTestBase {
     assertThat(actualWatermark).isEqualTo(expectedWatermarkOutput)
     assertThat(sink.getAppendResults.sorted).isEqualTo(expectedData.sorted)
   }
+
+  @Test
+  def testWatermarkNotMovingBack(): Unit = {
+    val data = Seq(
+      row(1, LocalDateTime.parse("2024-01-01T00:00:00")),
+      row(3, LocalDateTime.parse("2024-01-03T00:00:00")),
+      row(2, LocalDateTime.parse("2024-01-02T00:00:00"))
+    )
+
+    val dataId = TestValuesTableFactory.registerData(data)
+
+    val ddl =
+      s"""
+         | CREATE Table VirtualTable (
+         |   a INT,
+         |   c TIMESTAMP(3),
+         |   WATERMARK FOR c as c
+         | ) with (
+         |   'connector' = 'values',
+         |   'bounded' = 'false',
+         |   'scan.watermark.emit.strategy' = 'on-periodic',
```

Comment:
is it guaranteed that our tests are deterministic? usually on-periodic leads to non-deterministic results.

Reply:
https://github.com/apache/flink/pull/25921#discussion_r1907090365